File: js/day06/main.ts
import fs from 'fs'

function solve_day_one(input: string): number {
    const [times, records] = input.split('\n').map(x =>
        x
            .split(' ')
            .map(n => parseInt(n))
            .filter(n => !isNaN(n))
    )

    let total = 0

    for (let i = 0; i < times.length; i++) {
        const [time, record] = [times[i], records[i]]
        let c = 0
        for (let n = 1; n < time; n++) {
            const distance = n * (time - n)
            if (distance > record) {
                c++
            }
        }

        total = total ? total * c : c
    }

    return total
}

function solve_day_two(input: string): number {
    const [time, record] = input.split('\n').map(x =>
        parseInt(
            x
                .split(' ')
                .filter(n => !isNaN(parseInt(n)))
                .join('')
        )
    )

    let total = 0

    let c = 0
    for (let n = 1; n < time; n++) {
        const distance = n * (time - n)
        if (distance > record) {
            total++
        }
    }

    return total
}

function main() {
    const data = fs.readFileSync(0, 'utf-8')
    const part_one_result = solve_day_one(String(data))
    const part_two_result = solve_day_two(String(data))
    console.log(`part one: ${part_one_result}`)
    console.log(`part two: ${part_two_result}`)
}

main()

File: go/cmd/day06/main.go
package main

import (
	"fmt"
	"io"
	"os"
	"regexp"
	"strconv"
	"strings"
)

func solve_part_one(input_text string) int {
	total := 0

	times := regexp.MustCompile(`\d{1,}`).FindAllString(strings.Split(input_text, "\n")[0], -1)
	records := regexp.MustCompile(`\d{1,}`).FindAllString(strings.Split(input_text, "\n")[1], -1)

	i := 0
	for i < len(times) {
		t, err := strconv.Atoi(times[i])
		if err != nil {
			panic(err)
		}
		r, err := strconv.Atoi(records[i])
		if err != nil {
			panic(err)
		}

		n := 1
		c := 0
		for n < t {
			d := n * (t - n)
			if d > r {
				c++
			}

			n++
		}

		if total == 0 {
			total = c
		} else {
			total = total * c
		}

		i++

	}

	return total
}

func solve_part_two(input_text string) int {
	total := 0

	t, err := strconv.Atoi(strings.Split(strings.ReplaceAll(strings.Split(input_text, "\n")[0], " ", ""), ":")[1])
	if err != nil {
		panic(err)
	}
	r, err := strconv.Atoi(strings.Split(strings.ReplaceAll(strings.Split(input_text, "\n")[1], " ", ""), ":")[1])
	if err != nil {
		panic(err)
	}

	n := 1
	for n < t {
		d := n * (t - n)
		if d > r {
			total++
		}

		n++
	}

	return total
}

func main() {
	stdin, err := io.ReadAll(os.Stdin)
	if err != nil {
		panic(err)
	}

	input := string(stdin)

	part_one_result := solve_part_one(input)
	part_two_result := solve_part_two(input)
	fmt.Printf("part one: %d\n", part_one_result)
	fmt.Printf("part two: %d\n", part_two_result)
}
